### ISSUE TITLE
Use simple dynamic array instead of TArray::GetData() to avoid memory got recycled

### ISSUE THREAD

**Issue by ernestcheung** [carla-simulator/carla#543]
I was trying to additional attributes to Walker and try to send back to Python Client side, and I found a style of code that (probably) should be avoided).

In FCarlaServer::SendSensorData(const FSensorDataView&), the way that it encode a list of Non Player Agent into a repeated carla_agent is to create a TArray, fill it, and use TArray::GetData() to obtain a pointer to pass along to carla_sensor_data.

I followed this example to write the Encode function for my repeated attribute, however, the received values on client side are rewritten by the value of the last agent.  After a long way of debug (that lengthy compile time), I found out that obtaining a pointer from TArray::GetData() is not the best practice as the Unreal GC system that manages TArray is not aware of the presence of the reference to this pointer, and will reuse this piece of memory as it goes out of scope.  

This doesn't happen to SendSensorData for some reasons may be it just called for once in a tick, and after a tick.  My function is called for each walker and therefore that piece of memory has been keep recycled and rewritten, resulting in wrong data. 

In short, use dynamic array directly (carla_agent AgentsData = new calra_agent[N];) when creating new repeated attributes instead of TArray::GetData().  I suggest Carla make this modification so other developers reading this example code would not suffer the same issue again.

**Comment by nsubiron** [comment]
Hi @ernestcheung,

That's intentional. `SendSensorData` gets a "view" over a buffer (`FSensorDataView`), i.e. doesn't own the data, it's responsibility of the caller to make sure the buffer lives as long as this function call (just as `boost::asio::buffer` or `std::string_view`).

It's definitively riskier than passing a dynamic array, but the main reason to have this here is performance. This function is used to send images every frame, and making a copy of an image to a dynamic array is costly (a 1080 image is about 8MB, and we can have several of them each frame). Using this pattern we can directly pass the render buffer to this function. This function then passes it to the "CarlaServer" library that copies it to a pre-allocated buffer (double-buffer actually); so no dynamic memory is allocated during game.

Passing data with `TArray::GetData()` is totally fine as long as the array is not destroyed before `SendSensorData()` returns.

Now, the other reason to have these raw pointer to arrays is because "CarlaServer" is wrapped in a pure C header. The problem you're explaining happened to us too, when we send the measurements we send an array of `carla_agent`, this array has to be internally copied before the `carla_write_measurements()` function returns, that's why we have the class [CarlaMeasurements](https://github.com/carla-simulator/carla/blob/master/Util/CarlaServer/source/carla/server/CarlaMeasurements.h). This one reuses the buffer too to avoid dynamic allocation.

Hope it clarifies why we use this pattern.

Cheers!
